pkg/cloud/azure/services/availabilitysets/availabilitysets.go

```diff
@@ -0,0 +1,94 @@
+package availabilitysets
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-30/compute"
+	"github.com/Azure/go-autorest/autorest/to"
+	"sigs.k8s.io/cluster-api-provider-azure/pkg/cloud/azure"
+)
+
+// Spec input specification for Get/CreateOrUpdate/Delete calls
+type Spec struct {
+	Name string
+}
+
+func (s *Service) CreateOrUpdate(ctx context.Context, spec azure.Spec) error {
+	availabilitysetsSpec, ok := spec.(*Spec)
+	if !ok {
+		return errors.New("invalid availability set specification")
+	}
+
+	asParams := compute.AvailabilitySet{
+		Name: to.StringPtr(availabilitysetsSpec.Name),
+		Sku: &compute.Sku{
+			Name: to.StringPtr(string(compute.Aligned)),
+		},
+		Location: to.StringPtr(s.Scope.Location()),
+		// Todo: figure out if we need to set the tags
+		// Tags: ,
+		AvailabilitySetProperties: &compute.AvailabilitySetProperties{
+			PlatformFaultDomainCount:  to.Int32Ptr(int32(2)),
+			PlatformUpdateDomainCount: to.Int32Ptr(int32(5)),
+		},
+	}
+
+	_, err := s.Client.CreateOrUpdate(ctx, s.Scope.MachineConfig.ResourceGroup, availabilitysetsSpec.Name, asParams)
+	if err != nil {
+		return fmt.Errorf("failed to create availability set %s: %w", availabilitysetsSpec.Name, err)
+	}
+
+	return nil
+}
+
+// Get returns the availability set with the given name.
+func (s *Service) Get(ctx context.Context, spec azure.Spec) (interface{}, error) {
+	availabilitysetsSpec, ok := spec.(*Spec)
+	if !ok {
+		return nil, errors.New("invalid availability set specification")
+	}
+
+	as, err := s.Client.Get(ctx, s.Scope.MachineConfig.ResourceGroup, availabilitysetsSpec.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get availability set %s: %w", availabilitysetsSpec.Name, err)
+	}
+
+	return as, nil
+}
+
+// Delete deletes the availability set with the given name.
+func (s *Service) Delete(ctx context.Context, spec azure.Spec) error {
```

Comment:
Perhaps we should rename this to `DeleteIfEmpty` to make it more obvious?

Reply:
This function implements service interface so it can't be renamed.

Reply:
Ok, in that case I would suggest adding some more detail both in the comment on the function but also where it is called to mention that it will delete the AvailabilitySet only if it is empty. An alternative way to make this abundantly obvious would be to return an `ErrAvailabiltySetNotEmpty` error and have that checked at the call site